File: sessions/2018/medium/199_binary_tree_right_side_view.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        result = []

        if not root:
            return result

        q = [(root, 1)]
        level = 0
        while len(q) > 0:
            node, l = q.pop()
            if l > level:
                result.append(node.val)
                level = l
            if node.left:
                q.append((node.left, l + 1))
            if node.right:
                q.append((node.right, l + 1))
        return result

File: pia/2020/medium/199_binary_tree_right_side_view.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var rightSideView = function (root) {
  if (!root) {
    return [];
  }
  let list = [root];
  let visible = [];

  while (list.length) {
    const length = list.length;
    let temp = [];


    for (let i = 0; i < length; i++) {
      const node = list.shift();
      temp.push(node.val);

      if (node.left) {
        list.push(node.left);
      }
      if (node.right) {
        list.push(node.right);
      }
    }
    visible.push(temp[temp.length - 1]);
  }
  return visible;
};
